### PR TITLE
Adding recursive directory traversal functionality

### PR DESCRIPTION
**Pull request by jedi22**
Working pull request for adding the ability to specify queries with wildcards in them for recursive traversal and wildcard matching.

### CODE REVIEW

File: osquery/filesystem/filesystem.cpp [hunk 1]
```diff
@@ -147,6 +147,136 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+      }
+    }
+
+    return Status(0, "OK");
+  } catch (const boost::filesystem::filesystem_error& e) {
+    return Status(1, e.what());
+  }
+}
+
+std::vector<std::string> &split(const std::string &s, char delim, std::vector<std::string> &elems) {
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: https://github.com/facebook/osquery/blob/master/osquery/core/text.cpp#L18

File: osquery/filesystem/filesystem.cpp [hunk 2]
```diff
@@ -147,6 +147,136 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+    std::stringstream ss(s);
+    std::string item;
+    while (std::getline(ss, item, delim)) {
+      if(item.length() == 0){continue;}
+      elems.push_back(item);
+    }
+    return elems;
+}
+
+std::string crush(int start, int end, std::vector<std::string> components){
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: Use `boost::algorithm::join`: http://stackoverflow.com/questions/1833447/a-good-example-for-boostalgorithmjoin

File: osquery/filesystem/filesystem.cpp [hunk 3]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
   }
 }
 
+Status listDirectoriesInDirectory(const boost::filesystem::path& path,
+                            std::vector<std::string>& results) {
+  try {
+    if (!boost::filesystem::exists(path)) {
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: https://github.com/facebook/osquery/blob/master/osquery/filesystem/filesystem.cpp#L110

And you can move this out of a try.

File: osquery/filesystem/filesystem.cpp [hunk 4]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+      return Status(1, "Directory not found");
+    }
+
+    if (!boost::filesystem::is_directory(path)) {
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: https://github.com/facebook/osquery/blob/master/osquery/filesystem/filesystem.cpp#L159

Then move this out of a try.

File: osquery/filesystem/filesystem.cpp [hunk 5]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+    }
+  }
+
+  if(boost::filesystem::is_regular_file(fs_path)){
+    results.push_back(fs_path.string());
+  }
+  return Status(0, "OK");
+}
+/************************************************
+Outstanding Issues:
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: Don't leave this here, prefer to place in the commit message or the PR.

File: osquery/filesystem/filesystem.cpp [hunk 6]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+  Letters Star - Not implemented
+  Star Letters - Not implemented
+  Star Folders - Will be hard??
+  Imply Last * - Given /Users/* /Downloads/ should that be the same as /Users/* /Downloads/* ?
+  A Trailing / - Still thinks it's a file
+*************************************************/
+//Should be modified to already pass in the split path so that it doesn't have to be split everytime?
+Status listFilesInDirectoryWildRec(const boost::filesystem::path& fs_path, std::vector<std::string>& results, std::vector<std::string> components, unsigned int processed_index, unsigned int rec_depth){
+  //Handle all parts of the path except last because then we want to get files, not directories
+  for(unsigned int i = processed_index; i < components.size()-1; i++){
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: size_t works too

File: include/osquery/filesystem.h [hunk 7]
```diff
@@ -21,6 +21,8 @@
 
 namespace osquery {
 
+const unsigned int MAX_DEPTH = 40;
```
Comment (theopolis) [inline include/osquery/filesystem.h]: Maybe something more descriptive?

File: osquery/filesystem/filesystem.cpp [hunk 8]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+  } catch (const boost::filesystem::filesystem_error& e) {
+    return Status(1, e.what());
+  }
+}
+
+Status doFileListing(const boost::filesystem::path& fs_path, std::vector<std::string>& results, std::vector<std::string>& components){
+  if(components[components.size()-1][0] == '*' && components[components.size()-1].length() == 1){
+    std::vector<std::string> files;
+    Status stat = listFilesInDirectory(fs_path.parent_path(), files);
+    if(stat.getCode() != 0){return stat;}
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: can do `if (stat.ok()) {}` too

File: osquery/filesystem/filesystem.cpp [hunk 9]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+    for(auto& file : files){
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: `const auto&`?

File: osquery/filesystem/filesystem.cpp (hunk 6 above)
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: auto

File: include/osquery/filesystem.h (hunk 7 above)
Comment (marpaia) [inline include/osquery/filesystem.h]: kMaxDepth

File: osquery/filesystem/filesystem.cpp [hunk 10]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+        results.push_back(file);
+    }
+    return Status(0, "OK");
+  }
+  //If we get here, that means we've reached the last component of the original path.
+  //Check to make sure the path we've arrived at is valid at all.
+  try{
+    if(!boost::filesystem::exists(fs_path)){return Status(0, "OK");}
+  }catch(boost::filesystem::filesystem_error error){
+    VLOG(1) << "Permission was denied to access: " << fs_path.string();
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: This ends up being VERY VERY loud. I would recommend removing it or using `VLOG(3)`.

File: osquery/filesystem/filesystem.cpp [hunk 11]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+    return Status(0, "OK");
+  }
+  if(boost::filesystem::is_directory(fs_path)){
+    std::vector<std::string> files;
+    Status stat = listFilesInDirectory(fs_path, files);
+    if(stat.getCode() != 0){return stat;}
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: `if (stat.ok()) {}`

File: include/osquery/filesystem.h (hunk 7 above)
Comment (marpaia) [inline include/osquery/filesystem.h]: Our "style" with constants is kCamelCase. All uppercase is reserved for macros.

File: osquery/filesystem/filesystem.cpp [hunk 12]
```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+
+        //If had an error, stop and back out
+        if(stat.getCode() != 0)
+          return stat;{
+        }
+      }
+      // Our subcalls that handle processing are now complete, return
+      return Status(0, "OK");
+    }else{
+      //In these else statements is where we will handle letter star or star letter
```
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp]: Does this need to be implemented?

File: osquery/filesystem/filesystem.cpp (hunk 11 above)
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: `if (!stat.ok()) {`

Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: Also, never put entire if statements on one line, always break it up onto three:

```cpp
if (stat.ok()) {
  return stat;
}
```

**Comment by theopolis** [comment]
```python
print "PLEASE "*1000 + "use clang-format"
```

File: osquery/filesystem/filesystem.cpp (hunk 11 above)
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: I think clang-format will take care of that for you. Please clang-format so we don't have religious formatting discussion on PRs.

File: include/osquery/filesystem.h [hunk 13]
```diff
@@ -79,6 +85,8 @@ Status pathExists(const boost::filesystem::path& path);
 Status listFilesInDirectory(const boost::filesystem::path& path,
                             std::vector<std::string>& results);
 
+Status listFilesInDirectoryWild(const boost::filesystem::path& fs_path, std::vector<std::string>& results);
```
Comment (marpaia) [inline include/osquery/filesystem.h]: clang-format

File: osquery/filesystem/filesystem.cpp [hunk 14]
```diff
@@ -147,6 +148,175 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
   }
 }
 
+Status listDirectoriesInDirectory(const boost::filesystem::path& path,
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: This should probably be in `<osquery/filesystem.h>`, like `listFilesInDirectory`

File: osquery/filesystem/filesystem.cpp [hunk 15]
```diff
@@ -146,6 +148,181 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
...
+    } else if (components[i] == kWildcardCharacterRecursive) {
+      return Status(1, "** NOT LAST COMPONENT");
+    } else {
+      // When we want to support wildcards like "foobar%", this is where it
+      // will happen
+    }
+  }
+  // At this point, all of our call paths have been resolved, so know we want to
+  // list the files at this point or do our ** traversal
+  if (components[components.size() - 1] == kWildcardCharacterRecursive) {
```
Comment (marpaia) [inline osquery/filesystem/filesystem.cpp]: This will crash if `components.size()` == 0.

**Comment by jedi22** [comment]
test this please

**Comment by marpaia** [comment]
test this please

**Comment by jedi22** [comment]
test this please

**Pull request merged**